torchgen/api/types_base.py

```diff
@@ -0,0 +1,187 @@
+# (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.
+from abc import ABC
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional, Union
+from torchgen.model import (
+    Argument,
+    SelfArgument,
+    TensorOptionsArguments,
+)
+
+# An ArgName is just the str name of the argument in schema;
+# but in some special circumstances, we may add a little extra
+# context.  The Enum SpecialArgName covers all of these cases;
+# grep for their construction sites to see when they can occr.
+
+SpecialArgName = Enum("SpecialArgName", ("possibly_redundant_memory_format",))
+ArgName = Union[str, SpecialArgName]
+
+# This class shouldn't be created directly; instead, use/create one of the singletons below.
+@dataclass(frozen=True)
+class BaseCppType:
+    ns: Optional[str]
+    name: str
+
+    def __str__(self) -> str:
+        if self.ns is None or self.ns == "":
+            return self.name
+        return f"{self.ns}::{self.name}"
+
+class CType(ABC):
+    def cpp_type(self, *, strip_ref: bool = False) -> str:
+        raise NotImplementedError
+
+    def cpp_type_registration_declarations(self) -> str:
+        raise NotImplementedError
```

Comment:
The comments explaining what these methods mean need to get to the ABC, so people understand exactly what they're supposed to implement for these functions. But I'm also not a big fan of `cpp_type_registration_declarations`; the preexisting comment says that it's supposed to be for BC, can we get rid of it?

Reply:
I can do it in a separate PR